relayer/chain/ethereum/syncer/cache.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2020 Snowfork
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package syncer
+
+import (
+	gethCommon "github.com/ethereum/go-ethereum/common"
+	gethTypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+type HeaderCacheItem struct {
+	Header    *gethTypes.Header
+	Forwarded bool
+}
+
+// This is used to store the latest headers as they are published. Up to
+// `numHeightsToTrack` heights are stored. Once this number is reached, an old
+// height is pruned each time a new height is added. The current stored height
+// range is given by [minHeight, maxHeight].
+type HeaderCache struct {
+	headers           map[string]*HeaderCacheItem
+	hashesByHeight    map[uint64][]string
+	maxHeight         uint64
+	minHeight         uint64
+	numHeightsToTrack uint64
```

Comment:
> Doesn't that imply that if we are more than numHeightsToTrack behind the current tip of the blockchain, the cache will fail?

Yes and no. `numHeightsToTrack = descendantsUntilFinal + 1` in `NewSyncer`. So, provided that this cache is only used to retrieve headers after we've synced old blocks, it shouldn't matter that it doesn't contain older heights. After syncing old blocks, we will be `descendantsUntilFinal` behind the current tip

> There will be new blocks coming in at the top, but we try to sync old blocks at the beginning. It might make more sense to instead use a LRU cache, where we just store pointers in an array, pushing recently used ones to the top?

I'm not sure I understand this. We don't store old blocks in this cache. It's exclusively used for new headers as they come in. And it is a type of LRU cache in that we keep the latest heights and discard old heights.